### PR TITLE
NodeLifecycleController: Remove race condition

### PR DESCRIPTION
**Pull request by jprzychodzen**
Patch request does not support RV, and patching lists actually overwrites whole field. It means that there is a race condition, in which we can overwrite changes to taints that happened between GET and PATCH requests.

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

There is a race condition between KCM and CCM during cluster initialization CCM's Cloud Node controller removes `node.cloudprovider.kubernetes.io/uninitialized` taints, however KCM can add this taint do to race condition.

Issue was discovered during scale testing of a cluster with CCM by @mborsz .

#### Does this PR introduce a user-facing change?

 ```release-note
fix race condition between KCM's Node Lifecycle Controller and CCM's Cloud Node Controller
 ```

### CODE REVIEW

**Comment by jprzychodzen** [comment]
/hold

Add tests

File: pkg/controller/controller_utils.go [hunk 1]
```diff
@@ -1064,7 +1064,8 @@ func AddOrUpdateTaintOnNode(ctx context.Context, c clientset.Interface, nodeName
 		if !updated {
 			return nil
 		}
-		return PatchNodeTaints(ctx, c, nodeName, oldNode, newNode)
+		_, err = c.CoreV1().Nodes().Update(ctx, newNode, metav1.UpdateOptions{})
```
Comment (andrewsykim) [inline pkg/controller/controller_utils.go:1067]: For historical reasons there is a copy of `AddOrUpdateTaintOnNode` [here](https://github.com/jprzychodzen/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/node/helpers/taints.go#L51), should that also be updated or does only updating KCM fix the race condition?

Comment (jprzychodzen) [reply pkg/controller/controller_utils.go:1067]: This specific race condition happens because of KCM.

I prefer to have preemptive fix in other PR, as it's more a potential bug without materialized impact.

**Comment by jprzychodzen** [comment]
/unhold

Added test

**Comment by mborsz** [comment]
Please fix the presubmits

**Comment by andrewsykim** [comment]
@jprzychodzen summarizing my understanding of this race condition from our chat offline for posterity, please correct me if anything is inaccurate: 
 
The race condition is happening here because resource version validation for Patch checks against the existing object in etcd directly (see [updatedObjectInfo](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L662) and [objInfo.UpdatedObject](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L554-L558), when a nil object is passed to `updatedObjectInfo`, the returned updated object from `UpdatedObject` returns a resource version from the existing object). So when CCM does a patch in between when the KCM fetches the existing object from storage and when it updates the new object, it would erase the patch done by CCM, because the [validation for RVs](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L637-L639) is checking against a now stale version.

**Comment by jprzychodzen** [comment]
Generally, yes, however I would reword second part to

> So when CCM does an update in between when the KCM fetches the existing object from storage and when it updates the new object, it would erase the update done by CCM, because the [validation for RVs](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L637-L639) is using two time the same RV - from the same current object, unseen by KCM.

**Comment by liggitt** [comment]
> Patch request does not support RV

you can set metadata.resourceVersion in your patch if you want, and the API server will reject application of the patch if the RV is not up to date, but that means you will have to retry on conflict client-side

File: pkg/controller/controller_utils.go (hunk 1 above)
Comment (liggitt) [reply pkg/controller/controller_utils.go:1067]: in general, we try to avoid get/modify/write of the Node object because it means this will drop any Node fields unknown to this binary if it is linked to an older Node type definition

Comment (liggitt) [reply pkg/controller/controller_utils.go:1067]: that's because the patch still applies cleanly (does not conflict) even on top of the newly updated fields

**Comment by jprzychodzen** [comment]
> metadata.resourceVersion in your patch if you want

I was thinking about that, but either I can do this with [1] changing how strategicpatch.CreateTwoWayMergePatch works (to always include RV), [2] force add RV by removing RV from base node before generating patch with strategicpatch.CreateTwoWayMergePatch or [3] adding RV postfactum

I would like to go with version [2] if there is a strong standing that 'Update' is unsafe.

[1] would be api change, which is something that requires much more consideration - and I do na want to fix whole Patch request to resolve this race condition
[3] is hacky and seems fragile and error prone

**Comment by jprzychodzen** [comment]
/hold

To ensure that discussion around Update/Patch has a consensus.

**Comment by liggitt** [comment]
> [1] changing how strategicpatch.CreateTwoWayMergePatch works (to always include RV)

We definitely don't want to do that. That would ripple out to many many callers who do not want RV-dependent behavior

**Comment by liggitt** [comment]
> patching lists actually overwrites whole field

That depends on the definition of the list... in this case, the `taints` list does not have a merge strategy / key set, so you're correct, patches do override the whole list. That's unfortunate, and not possible to change compatibly. Ideally, that field would have set `patchStrategy:"merge" patchMergeKey:"key"`

**Comment by jprzychodzen** [comment]
/unhold

As per discussion with Jordan - went with [2] approach and still using Patch.

File: pkg/controller/controller_utils.go [hunk 2]
```diff
@@ -1127,7 +1127,9 @@ func RemoveTaintOffNode(ctx context.Context, c clientset.Interface, nodeName str
 
 // PatchNodeTaints patches node's taints.
 func PatchNodeTaints(ctx context.Context, c clientset.Interface, nodeName string, oldNode *v1.Node, newNode *v1.Node) error {
-	oldData, err := json.Marshal(oldNode)
+	oldNodeNoRV := oldNode.DeepCopy()
+	oldNodeNoRV.ResourceVersion = ""
```
Comment (alculquicondor) [inline pkg/controller/controller_utils.go:1131]: add comment for why this is necessary. Also, a shallow copy should be enough.

Comment (jprzychodzen) [reply pkg/controller/controller_utils.go]: Done.

File: pkg/controller/controller_utils.go [hunk 3]
```diff
@@ -1140,7 +1135,15 @@ func PatchNodeTaints(ctx context.Context, c clientset.Interface, nodeName string
 		return fmt.Errorf("failed to marshal new node %#v for node %q: %v", newNodeClone, nodeName, err)
 	}
 
-	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1.Node{})
+	// Strip base diff node from RV to ensure that our Patch request will set RV to check for conflicts over .spec.taints.
```
Comment (andrewsykim) [inline pkg/controller/controller_utils.go:1138]: Can we specifically mention that this is needed because .spec.taints does not specify `patchMergeKey` and `patchStrategy` and adding them is no longer an option for compatibility reasons

File: pkg/controller/controller_utils.go [hunk 4]
```diff
@@ -1127,11 +1127,6 @@ func RemoveTaintOffNode(ctx context.Context, c clientset.Interface, nodeName str
 
 // PatchNodeTaints patches node's taints.
 func PatchNodeTaints(ctx context.Context, c clientset.Interface, nodeName string, oldNode *v1.Node, newNode *v1.Node) error {
```
Comment (andrewsykim) [inline pkg/controller/controller_utils.go:1129]: @jprzychodzen now that we've updated this to include RV in the patch request, is there a reason not to include the same update in CCM [here](https://github.com/jprzychodzen/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/node/helpers/taints.go#L91)? At least this way both components will not conflict with other components in the cluster that would try to set taints on the Node.

File: pkg/controller/controller_utils.go (hunk 3 above)
Comment (jprzychodzen) [reply pkg/controller/controller_utils.go:1138]: Done.

Also, added information about using other patch type and it's conflict with removal of a trait.

File: pkg/controller/controller_utils.go (hunk 4 above)
Comment (jprzychodzen) [reply pkg/controller/controller_utils.go:1129]: Done

File: staging/src/k8s.io/cloud-provider/node/helpers/taints.go [hunk 5]
```diff
@@ -102,7 +97,17 @@ func PatchNodeTaints(c clientset.Interface, nodeName string, oldNode *v1.Node, n
 		return fmt.Errorf("failed to marshal new node %#v for node %q: %v", newNodeClone, nodeName, err)
 	}
 
-	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1.Node{})
+	// Strip base diff node from RV to ensure that our Patch request will set RV to check for conflicts over .spec.taints.
+	// This is needed because .spec.taints does not specify patchMergeKey and patchStrategy and adding them is no longer an option for compatibility reasons.
+	// Using other Patch strategy works for adding new taints, however will not resolve problem with taint removal.
+	oldNodeNoRV := oldNode
+	oldNodeNoRV.ResourceVersion = ""
```
[the remark this answers is not in the document]
Comment (aojea) [reply staging/src/k8s.io/cloud-provider/node/helpers/taints.go:104]: yeah, same in the other file `pkg/controller/controller_utils.go`

[the remark this answers is not in the document]
Comment (jprzychodzen) [reply staging/src/k8s.io/cloud-provider/node/helpers/taints.go:104]: I was doing it, by Aldo pointed out that we don't need it because we are working on a single copy of the `oldNode`, from function which exits on return (see [taints.go](https://github.com/kubernetes/kubernetes/blob/32411016cc84d241558bc9738a525f5a8f7576d6/staging/src/k8s.io/cloud-provider/node/helpers/taints.go#L86))

[the remark this answers is not in the document]
Comment (jprzychodzen) [reply staging/src/k8s.io/cloud-provider/node/helpers/taints.go:96]: Moved back to deep copy, as it seems that performance gain is negligible and this might be a source of some funny bugs in the future.

File: pkg/controller/controller_utils.go (hunk 2 above)
Comment (jprzychodzen) [reply pkg/controller/controller_utils.go:1134]: Reverted, leaving as a deep copy to make sure that we do not modify object that was passed to us. It seems not worth the performance gain.

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

**Comment by EmilyShepherd** [comment]
I just ran into a similar issue - part of why this was a problem for me is that I was expecting the taint list to be `listType=map` so that multiple controllers could use server side apply without clobbering each others changes.

Is there a reason that the node taint list is currently set to atomic? This behaviour requires each controller to respect each others changes when forming patches, which helps lead to this race condition. It also makes SSA unusable unless always forcing taint changes.
